examples/alife-boid/src/boid.rs

```diff
@@ -0,0 +1,117 @@
+use crate::rand::Rand;
+use crate::vector::Vector;
+use std::f64::consts::PI;
+
+const HEIGHT: f64 = 400.0;
+const WIDTH: f64 = 600.0;
+const VELOCITY_SIZE: f64 = 5.0;
+const ALIGNMENT_RADIOUS: f64 = 100.0;
+const ALIGNMENT_WEIGHT: f64 = 3.0;
+const COHENSION_RADIOUS: f64 = 200.0;
+const COHENSION_WEIGHT: f64 = 1.0;
+const SEPARATION_RADIOUS: f64 = 50.0;
+const SEPARATION_WEIGHT: f64 = 1.0;
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct Boid {
+    pub position: Vector,
+    pub velocity: Vector,
+}
+
+impl Boid {
+    pub fn new(rng: &mut Rand) -> Boid {
+        let theta = rng.next_f64() * PI * 2.0;
+        Boid {
+            position: Vector::new(WIDTH * rng.next_f64(), HEIGHT * rng.next_f64()),
+            velocity: Vector::new(theta.cos() * VELOCITY_SIZE, theta.sin() * VELOCITY_SIZE),
+        }
+    }
+
+    fn calc_alignment(&self, boids: &[Boid]) -> Vector {
+        let mut ret = Vector::new(0.0, 0.0);
+        for other in boids {
+            let mut position = other.position.clone();
+            position.sub(&self.position);
+            let position_size = position.size();
+            if position_size == 0.0 || position_size > ALIGNMENT_RADIOUS {
+                continue;
+            }
+
+            /*
+            let mut velocity = other.velocity.clone();
+            velocity.sub(&self.velocity);
+            */
+            ret.add(&other.velocity);
+        }
+
+        ret.normalize();
+        ret.mul(ALIGNMENT_WEIGHT);
+        ret
+    }
+
+    fn calc_cohension(&self, boids: &[Boid]) -> Vector {
```

Comment:
This function calculates the average offset to each other boid for each boid.
Instead of doing it for each and every one, you can calculate the sum position once and re-use it for all boids.

Calculating the cohesion weight is then as cheap as this:

```rust
// POS_SUM is the pre-calculated sum of all positions
// calculate the average position of all boids except this one.
let avg_pos = (POS_SUM - self.position) / (boids.length() - 1);
// my offset from said average position
let mut offset = (self.position - avg_pos);
offset.normalize();
offset *= COHESION_WEIGHT;
offset
```

Reply:
Btw, usually the cohesion magnitude scales linearly with the distance but divided by a factor so it stays reasonably small:
```rust
offset /= COHESION_WEIGHT;
```